### PR TITLE
ENH: inset_axes anchored-artist API

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

`ax.inset_axes` is now merged for 3.0 (#11026) with a minimal positioning API.  (`ax.inset_axes([l, b, w, h])`), but folks can specify transforms that allow a pad with the edge of the axes.

This PR expands the API to be similar to `ax.legend` and `axes_grid1.inset_axes` and allow a string for the position argument.  So now we can do `ax.inset_axes('NE', width=0.4, height=0.4)`

```python 
import matplotlib.pyplot as plt
import numpy as np


def get_demo_image():
    from matplotlib.cbook import get_sample_data
    import numpy as np
    f = get_sample_data("axes_grid/bivariate_normal.npy", asfileobj=False)
    z = np.load(f)
    Z2 = np.zeros([150, 150], dtype="d")
    ny, nx = z.shape
    Z2[30:30 + ny, 30:30 + nx] = z

    # z is a numpy array of 15x15
    extent = (-3, 4, -4, 3)
    return Z2, extent

fig, ax = plt.subplots(figsize=[5, 4])

# make data
Z2, extent = get_demo_image()
ax.imshow(Z2, extent=extent, interpolation="nearest",
          origin="lower")

# inset axes....
axins = ax.inset_axes('NE', width=0.4, height=0.4)

axins.imshow(Z2, extent=extent, interpolation="nearest",
          origin="lower")
# sub region of the original image
x1, x2, y1, y2 = -1.5, -0.9, -2.5, -1.9
axins.set_xlim(x1, x2)
axins.set_ylim(y1, y2)
axins.set_xticklabels(''); axins.set_yticklabels('')

ax.indicate_inset_zoom(axins)
fig.canvas.draw()
plt.show()
```

boo

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is PEP 8 compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by jklymak** [comment]
ping @ImportanceOfBeingErnest  who's concern about the API sparked this.

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -431,13 +503,28 @@ def inset_axes(self, bounds, *, transform=None, zorder=5,
...
 
-        bounds : [x0, y0, width, height]
-            Lower-left corner of inset axes, and its width and height.
+        bounds : [x0, y0, width, height] or string.
+            If four-tupple: lower-left corner of inset axes, and its width and
+            height.
+
+            If a string, then locations such as "NE", "N", "NW", "W", etc,
+            of "upper right", "top", "upper left", etc (see `~.axes.legend`)
+            for codes.  (Note we do *not* support the numerical codes).
```
Comment (ImportanceOfBeingErnest) [inline lib/matplotlib/axes/_axes.py]: As long as other functions  like `legend` support the number codes, I would support them everywhere. Once it is decided to deprectate them, I would deprecate them all, everywhere, in one single rush. One might silently allow them here, without mentionning them specifically though. <- my personal opinion (I'm not too inclined to open that discussion again as it will just never get to an end.)

File: lib/matplotlib/backends/backend_qt5agg.py [hunk 2]
```diff
@@ -48,10 +48,10 @@ def paintEvent(self, event):
             [[left, self.renderer.height - (top + height * self._dpi_ratio)],
              [left + width * self._dpi_ratio, self.renderer.height - top]])
         reg = self.copy_from_bbox(bbox)
-        buf = reg.to_string_argb()
+        # buf = reg.to_string_argb()
+        buf = memoryview(reg)
```
Comment (ImportanceOfBeingErnest) [inline lib/matplotlib/backends/backend_qt5agg.py]: I'm almost certain that this should not be part of this PR.

File: lib/matplotlib/tests/test_axes.py [hunk 3]
```diff
@@ -5782,3 +5782,46 @@ def test_zoom_inset():
                    [0.8425, 0.907692]])
     np.testing.assert_allclose(axin1.get_position().get_points(),
             xx, rtol=1e-4)
+
+
+def test_inset_codes():
+    "Test that the inset codes put the inset where we want..."
```
Comment (ImportanceOfBeingErnest) [inline lib/matplotlib/tests/test_axes.py]: Single double-quotes as function docs are pretty non-standard. Not sure if it matters for tests though.

File: lib/matplotlib/backends/backend_qt5agg.py (hunk 2 above)
Comment (jklymak) [reply lib/matplotlib/backends/backend_qt5agg.py]: Ha, no...

File: lib/matplotlib/tests/test_axes.py [hunk 4]
```diff
@@ -5782,3 +5782,46 @@ def test_zoom_inset():
...
+            [[0.135417, 0.113889],
+             [0.329167, 0.313889]],
+            [[0.135417, 0.4     ],
+             [0.329167, 0.6     ]],
+            [[0.135417, 0.686111],
+             [0.329167, 0.886111]]]
+    codes = ['N', 'NE', 'E', 'SE', 'S', 'SW', 'W', 'NW']
+    for pos, code in zip(poss, codes):
+        axin1 = ax.inset_axes(code)
+        np.testing.assert_allclose(axin1.get_position().get_points(),
```
Comment (ImportanceOfBeingErnest) [inline lib/matplotlib/tests/test_axes.py]: Would it make sense to define some `bbox`es in axes coordinates here and compare to their coordinates instead? (Seems like this test is pretty susceptible to unrelated changes in the subplot positioning.)

File: lib/matplotlib/axes/_axes.py [hunk 5]
```diff
@@ -110,6 +110,77 @@ def inset_locator(ax, renderer):
...
+             'lower right':  'SE',
+             'right':        'E',
+             'left':         'W',
+             'bottom':       'S',
+             'top':          'N',
+             'center left':  'W',
+             'center right': 'E',
+             'lower center': 'S',
+             'upper center': 'N',
+             }
```
Comment (ImportanceOfBeingErnest) [inline lib/matplotlib/axes/_axes.py]: What about the center itself? I.e. loc code `10`.

File: lib/matplotlib/axes/_axes.py [hunk 6]
```diff
@@ -431,13 +503,28 @@ def inset_axes(self, bounds, *, transform=None, zorder=5,
...
         ----------
 
-        bounds : [x0, y0, width, height]
-            Lower-left corner of inset axes, and its width and height.
+        bounds : [x0, y0, width, height] or string.
+            If four-tupple: lower-left corner of inset axes, and its width and
+            height.
+
+            If a string, then locations such as "NE", "N", "NW", "W", etc,
+            of "upper right", "top", "upper left", etc (see `~.axes.legend`)
```
[the remark this answers is not in the document]
Comment (ImportanceOfBeingErnest) [reply lib/matplotlib/axes/_axes.py]: And if "center" is unsopported (see above), one cannot simply link to `axes.legend`.

**Comment by jklymak** [comment]
> something like the mentionned `add_inset_complete` would be needed.

I *think* this has all the features of `add_inset_complete`.  Note the `transform` kwarg works with the new API and the `borderaxespad` parameter gives the padding.  You can also specify a `bbox_to_anchor`.  The difference is that we need to specify `width` and `height` or take the defaults, which you don't need to do for legend because it shrinks to fit.

File: lib/matplotlib/axes/_axes.py (hunk 1 above)
Comment (jklymak) [reply lib/matplotlib/axes/_axes.py]: I hear you, but someone can add that as a followup PR if they really feel strongly about it.  Its a bunch of extra code for something we don't really want to support any longer.

File: lib/matplotlib/tests/test_axes.py (hunk 4 above)
Comment (jklymak) [reply lib/matplotlib/tests/test_axes.py]: The string-code positioning has a borderpad that is in pixels, not axes co-ordinates, so knowing the bbox a-priori would just mean I was duplicating the code in the method.  I guess that'd test if someone changed it ...

File: lib/matplotlib/axes/_axes.py (hunk 5 above)
Comment (jklymak) [reply lib/matplotlib/axes/_axes.py]: Done.  Thanks for all the help!

**Comment by ImportanceOfBeingErnest** [comment]
I went through [the inset_locator_demo](https://matplotlib.org/devdocs/gallery/axes_grid1/inset_locator_demo.html#sphx-glr-gallery-axes-grid1-inset-locator-demo-py) and indeed almost all cases are feasible with the proposed API.

<details>
 <summary>Here is the changed example code</summary>
 
```
import matplotlib.pyplot as plt
#from mpl_toolkits.axes_grid1.inset_locator import inset_axes
#
fig, (ax, ax2) = plt.subplots(1, 2, figsize=[5.5, 2.8])

# Create inset of width 1.3 inches and height 0.9 inches
# at the default upper right location
#axins = inset_axes(ax, width=1.3, height=0.9)
axins = ax.inset_axes("NE", width=1.3, height=0.9, transform=fig.dpi_scale_trans)

# Create inset of width 30% and height 40% of the parent axes' bounding box
# at the lower left corner (loc=3)
#axins2 = inset_axes(ax, width="30%", height="40%", loc=3)
axins2 = ax.inset_axes("SW", width=0.3, height=.4)

# Create inset of mixed specifications in the second subplot;
# width is 30% of parent axes' bounding box and
# height is 1 inch at the upper left corner (loc=2)
#axins3 = inset_axes(ax2, width="30%", height=1., loc=2)
from matplotlib.transforms import blended_transform_factory
trans = blended_transform_factory(ax2.transAxes, fig.dpi_scale_trans)
axins3 = ax2.inset_axes("NW", width=0.3, height=1., transform=trans)

# Create an inset in the lower right corner (loc=4) with borderpad=1, i.e.
# 10 points padding (as 10pt is the default fontsize) to the parent axes
#axins4 = inset_axes(ax2, width="20%", height="20%", loc=4, borderpad=1)
axins4 = ax2.inset_axes("SE", width=.2, height=.2, borderaxespad=10)

# Turn ticklabels of insets off
for axi in [axins, axins2, axins3, axins4]:
    axi.tick_params(labelleft=False, labelbottom=False)

plt.show()

## =============================================================================
##
from matplotlib import transforms as mtrans
fig = plt.figure(figsize=[5.5, 2.8])
ax = fig.add_subplot(121)

# We use the axes transform as bbox_transform. Therefore the bounding box
# needs to be specified in axes coordinates ((0,0) is the lower left corner
# of the axes, (1,1) is the upper right corner).
# The bounding box (.2, .4, .6, .5) starts at (.2,.4) and ranges to (.8,.9)
# in those coordinates.
# Inside of this bounding box an inset of half the bounding box' width and
# three quarters of the bounding box' height is created. The lower left corner
# of the inset is aligned to the lower left corner of the bounding box (loc=3).
# The inset is then offset by the default 0.5 in units of the font size.

#axins = inset_axes(ax, width="50%", height="75%",
#                   bbox_to_anchor=(.2, .4, .6, .5),
#                   bbox_transform=ax.transAxes, loc=3)
# This is hard to reproduce, because padding is applied towards the axes
# There is no bounding box.
trans = ax.transAxes  + mtrans.ScaledTranslation(+5/72., +5/72, fig.dpi_scale_trans) 
axins = ax.inset_axes((.2, .4, .6*0.5, .5*0.75), transform=trans)


# For visualization purposes we mark the bounding box by a rectangle
ax.add_patch(plt.Rectangle((.2, .4), .6, .5, ls="--", ec="c", fc="None",
                           transform=ax.transAxes))

# We set the axis limits to something other than the default, in order to not
# distract from the fact that axes coodinates are used here.
ax.axis([0, 10, 0, 10])


# Note how the two following insets are created at the same positions, one by
# use of the default parent axes' bbox and the other via a bbox in axes
# coordinates and the respective transform.
ax2 = fig.add_subplot(222)
#axins2 = inset_axes(ax2, width="30%", height="50%")
axins2 = ax2.inset_axes("NE", width=.3, height=.5)


ax3 = fig.add_subplot(224)
trans = ax3.transAxes  + mtrans.ScaledTranslation(-5/72., -5/72, fig.dpi_scale_trans) 
axins3 = ax3.inset_axes((.7, .5, .3, .5), width="100%", height="100%",
                        transform=trans)

# For visualization purposes we mark the bounding box by a rectangle
ax2.add_patch(plt.Rectangle((0, 0), 1, 1, ls="--", lw=2, ec="c", fc="None"))
ax3.add_patch(plt.Rectangle((.7, .5), .3, .5, ls="--", lw=2,
                            ec="c", fc="None"))

# Turn ticklabels off
for axi in [axins2, axins3, ax2, ax3]:
    axi.tick_params(labelleft=False, labelbottom=False)

plt.show()
#
#
## =============================================================================
##
#
#
fig = plt.figure(figsize=[5.5, 2.8])
ax = fig.add_subplot(131)

# Create an inset outside the axes
#axins = inset_axes(ax, width="100%", height="100%",
#                   bbox_to_anchor=(1.05, .6, .5, .4),
#                   bbox_transform=ax.transAxes, loc=2, borderpad=0)
axins = ax.inset_axes((1.05, .6, .5, .4))
axins.tick_params(left=False, right=True, labelleft=False, labelright=True)

# Create an inset with a 2-tuple bounding box. Note that this creates a
# bbox without extent. This hence only makes sense when specifying
# width and height in absolute units (inches).
#axins2 = inset_axes(ax, width=0.5, height=0.4,
#                    bbox_to_anchor=(0.33, 0.25),
#                    bbox_transform=ax.transAxes, loc=3, borderpad=0)
# -----------
#This seems impossible, even when using a 4 tuple bounds.
# -----------

ax2 = fig.add_subplot(133)
ax2.set_xscale("log")
ax2.axis([1e-6, 1e6, -2, 6])

# Create inset in data coordinates using ax.transData as transform
#axins3 = inset_axes(ax2, width="100%", height="100%",
#                    bbox_to_anchor=(1e-2, 2, 1e3, 3),
#                    bbox_transform=ax2.transData, loc=2, borderpad=0)
axins3 = ax2.inset_axes((1e-2, 2, 1e3, 3), transform=ax2.transData)


# Create an inset horizontally centered in figure coordinates and vertically
# bound to line up with the axes.
from matplotlib.transforms import blended_transform_factory
transform = blended_transform_factory(fig.transFigure, ax2.transAxes)
#axins4 = inset_axes(ax2, width="16%", height="34%",
#                    bbox_to_anchor=(0, 0, 1, 1),
#                    bbox_transform=transform, loc=8, borderpad=0)
axins4 = ax2.inset_axes((.5-.16/2, 0, .16, .34), transform=transform)
plt.show()
```

</details>

While some cases become much more straight forward, e.g.

```
axins = inset_axes(ax, width="100%", height="100%",
                   bbox_to_anchor=(1.05, .6, .5, .4),
                   bbox_transform=ax.transAxes, loc=2, borderpad=0)
```
becomes

    axins = ax.inset_axes((1.05, .6, .5, .4))

others are more intricate, e.g. 

    axins3 = inset_axes(ax2, width="30%", height=1., loc=2)

becomes:
```
from matplotlib.transforms import blended_transform_factory
trans = blended_transform_factory(ax2.transAxes, fig.dpi_scale_trans)
axins3 = ax2.inset_axes("NW", width=0.3, height=1., transform=trans)
```

Now there is one example, where I did not find any way to reproduce. This should create an 0.5 by 0.4 inch inset at position (0.33, 0.25) in axes coordinates. 

```
axins2 = inset_axes(ax, width=0.5, height=0.4,
                    bbox_to_anchor=(0.33, 0.25),
                    bbox_transform=ax.transAxes, loc=3, borderpad=0)
```

Despite this I would still approve this, as the API is in general much simplified in contrast to the axes_grid1 solution. Still, the above shows that the axes_grid1 method has a lot of benefits and is still more generally applicable.

**Comment by jklymak** [comment]
@ImportanceOfBeingErnest thanks for looking at this so carefully.  

There is no huge rush on this, so I will look at your cases and see if an API can be worked out that can cover those as well.  Moving back to WIP...
